llvm/lib/Transforms/IPO/FatLTOCleanup.cpp

```diff
@@ -0,0 +1,125 @@
+// correct for the non-LTO part of a FatLTO object.
+//
+// llvm.type.checked.load instruction sequences always have a particular form:
+//
+// clang-format off
+//
+//   %0 = tail call { ptr, i1 } @llvm.type.checked.load(ptr %vtable, i32 0, metadata !"foo"), !nosanitize !0
+//   %1 = extractvalue { ptr, i1 } %0, 1, !nosanitize !0
+//   br i1 %1, label %cont2, label %trap1, !nosanitize !0
+//
+// trap1:                                            ; preds = %entry
+//   tail call void @llvm.ubsantrap(i8 2) #3, !nosanitize !0
+//   unreachable, !nosanitize !0
+//
+// cont2:                                            ; preds = %entry
+//   %2 = extractvalue { ptr, i1 } %0, 0, !nosanitize !0
+//   %call = tail call noundef i64 %2(ptr noundef nonnull align 8 dereferenceable(8) %p1) #4
+//
+// clang-format on
+//
+// In this sequence, the vtable pointer is first loaded and checked against some
+// metadata. The result indicates failure, then the program traps. On the
+// success path, the pointer is used to make an indirect call to the function
+// pointer loaded from the vtable.
+//
+// Since we won't be able to lower this correctly later in non-LTO builds, we
+// need to drop the special load and trap, and emit a normal load of the
+// function pointer from the vtable.
+//
+// This is straight forward, since the checked load can be replaced w/ a load
+// of the vtable pointer and a GEP instruction to index into the vtable and get
+// the correct method/function pointer. We replace the "check" with a constant
+// indicating success, which allows later passes to simplify control flow and
+// remove any now dead instructions.
+//
+// This logic holds for both llvm.type.checked.load and
+// llvm.type.checked.load.relative instructions.
+static bool cleanUpTypeCheckedLoad(Module &M, Function &CheckedLoadFn,
+                                   bool IsRelative) {
+  bool Changed = false;
+  for (Use &U : llvm::make_early_inc_range(CheckedLoadFn.uses())) {
+    for (auto *User : U->users()) {
+      Instruction *I = dyn_cast<Instruction>(User);
+      if (!I)
+        continue;
+      IRBuilder<> IRB(I);
+      Value *Ptr = I->getOperand(0);
+      Value *Offset = I->getOperand(1);
+      Type *PtrTy = I->getType()->getStructElementType(0);
+      ConstantInt *True = ConstantInt::getTrue(M.getContext());
+      Instruction *Load;
+      if (IsRelative) {
+        Function *LoadRelIntrinsic = llvm::Intrinsic::getOrInsertDeclaration(
+            &M, Intrinsic::load_relative, {Offset->getType()});
+        Load = IRB.CreateCall(LoadRelIntrinsic, {Ptr, Offset}, "rel_load");
+      } else {
+        Value *PtrAdd = IRB.CreatePtrAdd(Ptr, Offset);
+        Load = IRB.CreateLoad(PtrTy, PtrAdd, "vfunc");
+      }
+
+      Value *Replacement = PoisonValue::get(I->getType());
+      Replacement = IRB.CreateInsertValue(Replacement, Load, {0});
+      Replacement = IRB.CreateInsertValue(Replacement, True, {1});
```

Comment:
If you swap these these, the generated code will be slightly simpler due to constant folding.

Reply:
Much nicer. Thanks. I'll try to keep that in mind.